### PR TITLE
SNOW-279791 Basics of streaming snowpipe 

### PR DESCRIPTION
**Pull request by sfc-gh-japatel**
This is a PR for pushing the already approved changes to master so that subsequent PRs can be opened against master. 

There is no change apart from merging these three together in this PR.
1. https://github.com/snowflakedb/snowflake-kafka-connector/pull/379
2. https://github.com/snowflakedb/snowflake-kafka-connector/pull/381
3. https://github.com/snowflakedb/snowflake-kafka-connector/pull/382

I have reverted one change which was using a built jar, instead we can use beta version of ingest SDK.

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+    TopicPartitionChannel topicPartitionChannel =
+        new TopicPartitionChannel(partitionChannel, this.conn, tableName);
+    partitionsToChannel.putIfAbsent(partitionChannelKey, topicPartitionChannel);
+  }
+
+  /**
+   * Inserts the given record into buffer and then eventually calls insertRows API if buffer
+   * threshold has reached.
+   *
+   * <p>TODO: SNOW-473896 - Please note we will get away with Buffering logic in future commits.
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:145]: I'm not sure about this, could we remove this comment for now?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+    if (!partitionsToChannel.containsKey(partitionChannelKey)) {
+      LOGGER.warn(
+          "Topic: {} Partition: {} hasn't been initialized by OPEN " + "function",
+          record.topic(),
+          record.kafkaPartition());
+      startTask(
+          Utils.tableName(record.topic(), this.topicToTableMap),
+          record.topic(),
+          record.kafkaPartition());
+    }
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:195]: so this is possible during rebalance?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+    partitions.forEach(
+        topicPartition -> {
+          String partitionChannelKey =
+              partitionChannelKey(topicPartition.topic(), topicPartition.partition());
+          LOGGER.info("Closing partition channel:{}", partitionChannelKey);
+          partitionsToChannel.get(partitionChannelKey).closeChannel();
+          partitionsToChannel.remove(partitionChannelKey);
+        });
+
+    closeStreamingClient();
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:258]: If we're going to close streaming client, why not just call `closeAll`?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+  }
+
+  @Override
+  public void setTopic2TableMap(Map<String, String> topicToTableMap) {
+    this.topicToTableMap = topicToTableMap;
+  }
+
+  @Override
+  public void setFlushTime(long time) {
+    if (time < SnowflakeSinkConnectorConfig.BUFFER_FLUSH_TIME_SEC_MIN) {
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:310]: I think we will need a separate min for SS? And we will need to update the flush interval logic if we're updating this to something like 1s, probably need a BG to check it?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+    }
+    return streamingIngestClient.openChannel(channelRequest);
+  }
+
+  /* Init Streaming client. If is also used to re-init the client if client was closed before. */
+  private void initStreamingClient() {
+    Properties streamingClientProps = new Properties();
+    streamingClientProps.putAll(connectorConfig);
+    if (this.streamingIngestClient == null || this.streamingIngestClient.isClosed()) {
+
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:405]: nit: extra space

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+      this.streamingIngestClient =
+          SnowflakeStreamingIngestClientFactory.builder(this.streamingIngestClientName)
+              .setProperties(streamingClientProps)
+              .build();
+    }
+  }
+
+  private void closeStreamingClient() {
+    // do we need to close the client? If I close, I will have to re init the client upon rebalance
+    // in open()
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:415]: please remove, i think the decision is we will close?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+  // Offset number we would want to commit back to kafka
+  // This value is + 1 of what we find in snowflake.
+  // It tells kafka to start sending offsets from this offset because earlier ones have been
+  // ingested. (Hence it +1)
+  private final AtomicLong committedOffset;
+
+  // added to buffer before calling insertRows
+  private final AtomicLong processedOffset; // processed offset
+
+  // Ctor
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:76]: javadoc comment please

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+            (Iterable<Map<String, Object>>) intermediateBuffer.getData(),
+            Long.toString(intermediateBuffer.getLastOffset()));
+
+    // need to update the flush time(last I called insertrows)
+    this.previousFlushTimeStampMs = System.currentTimeMillis();
+
+    // should we retry if response.hasErrors() == true?
+    // how can we retry if buffer was cleared?
+    // lets have an exponential back off and jitter retry policy
+
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:229]: I think we made the decision, could you update the comment with TODOs?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+                    insertError.getException().getMessage());
+              });
+    }
+  }
+
+  // should we rely on precommits frequency to give us the last committed offset token?
+  // or should we have a BG thread polling on every channel to update the Atomic long and simply
+  // return the atomic long's current value
+  // For now we will rely on preCommit API to fetch the inserted offsets into Snowflake
+  // TODO: SNOW-529755 POLL committed offsets in backgraound thread
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:251]: same here, please add TODO to the javadoc

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+    LOGGER.debug(
+        "Fetching last committed offset for partition channel:{}",
+        this.channel.getFullyQualifiedName());
+    String lastOffsetCommitted = channel.getLatestCommittedOffsetToken();
+    LOGGER.info(
+        "Last committed offset for partition channel:{} is :{}",
+        this.channel.getFullyQualifiedName(),
+        lastOffsetCommitted);
+    if (Strings.isNullOrEmpty(lastOffsetCommitted)) {
+      return committedOffset.get();
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:277]: Should we return -1 or 0 here?

**Review by sfc-gh-tzhang — APPROVED** [review]
Discussed offline, the comments could be addressed in a separate PR given that this is a pure merge from approved PRs. Let's go!

**Comment by sfc-gh-tzhang** [comment]
@sfc-gh-japatel please make sure the end to end testing is passing before merging, it's failing right now

**Comment by sfc-gh-japatel** [comment]
> Discussed offline, the comments could be addressed in a separate PR given that this is a pure merge from approved PRs. Let's go!

Thanks Toby, I created https://snowflakecomputing.atlassian.net/browse/SNOW-541747..

Re: Failed merge gate, the GCP gate is flaky because of slowness. If rest of the tests are passing, we should be good to go..

**Pull request merged**